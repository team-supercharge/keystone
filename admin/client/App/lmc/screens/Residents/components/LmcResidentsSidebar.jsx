import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { ActionCreators } from '../../../actions/actions'
import { GlyphButton, ResponsiveText } from '../../../../elemental'
import _ from 'lodash'
import LmcSidebarItem from '../../../components/LmcSidebarItem.jsx'
import LmcResidentsSidebarFilter from './LmcResidentsSidebarFilter.jsx'

export class LmcResidentsSidebar extends Component {
    constructor(props) {
        super(props)
    }

    state = {
        nameFilter: '',
        displayActiveResidents: true
    }

    componentDidMount () {
        const { residents, selectedResident, setSelectedResident } = this.props;
        let shownResidents = _.filter(residents, (resident) => !this.calculateHidden(resident))
        if (shownResidents.length && !selectedResident) {
            setSelectedResident(shownResidents[0].id)
        }
    }

    calculateHidden = (resident) => {
        return (
            !resident.name.match(new RegExp(this.state.nameFilter, 'i')) || 
            (this.state.displayActiveResidents && resident.fields.status !== 'active')
        )
    }

    handleFormChange = (e) => {
        this.setState({
            [e.target.name]: e.target.value
        })
    }

    handleSwitchChange = () => {
        this.setState(prevState => ({ 
            displayActiveResidents: !prevState.displayActiveResidents 
        }))
    }

    render() {
        const { 
            residents, 
            selectedResident, 
            setSelectedResident,
            onCreate,
        } = this.props;

        let shownResidents = _.filter(residents, (resident) => !this.calculateHidden(resident))

        return (
            <div className='lmc-sidebar' style={styles.container}>
                <LmcResidentsSidebarFilter
                    onFormChange={this.handleFormChange}
                    onSwitchChange={this.handleSwitchChange}
                    isChecked={!this.state.displayActiveResidents}
                />
                <GlyphButton
                    block
                    color='success'
                    glyph='plus'
                    position='left'
                    title={ADD_RESIDENTS_BUTTON_TEXT}
                    onClick={() => onCreate('Residents')}
                    style={styles.button}
                >
                    <ResponsiveText
                        visibleSM={ADD_RESIDENTS_BUTTON_TEXT}
                        visibleMD={ADD_RESIDENTS_BUTTON_TEXT}
                        visibleLG={ADD_RESIDENTS_BUTTON_TEXT}
                     />
                </GlyphButton>
                <ul className='lmc-sidebar-list'>
                    { shownResidents.map((resident, index) => {
                            return (
                                <LmcSidebarItem 
                                    key={index}
                                    itemData={resident}
                                    onClick={() => setSelectedResident(resident.id)}
                                    isSelected={resident.id === selectedResident}
                                />
                            )
                        }) }
                </ul>
            </div>
        )
    }
}

const styles = {
    button: {
        borderRadius: 0,
    },
    container: {
        flex: '1',
        zIndex: '1'
    }
};

const ADD_RESIDENTS_BUTTON_TEXT = 'Add a Resident';

LmcResidentsSidebar.propTypes = {
    residents: PropTypes.array,
    selectedResident: PropTypes.string,
    onCreate: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
    return {
        selectedResident: state.residents.selectedResident,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        setSelectedResident: (id) => dispatch(ActionCreators.setSelectedResident(id)),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LmcResidentsSidebar);