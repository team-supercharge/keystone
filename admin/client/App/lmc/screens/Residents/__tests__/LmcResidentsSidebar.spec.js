import React from 'react'
import { shallow } from 'enzyme'
import { LmcResidentsSidebar } from '../components/LmcResidentsSidebar.jsx'

describe('LmcResidentsSidebar', () => {
    let wrapper
    let residents
    const onCreateMock = jest.fn()

    beforeEach(() => {
        residents = [
            { id: 'testId1', name: { first: 'test', last: 'id1' }, status: 'active' },
            { id: 'testId2', name: { first: 'test', last: 'id2' }, status: 'active' },
            { id: 'testId3', name: { first: 'test', last: 'id3' }, status: 'inactive' },
        ]
        wrapper = shallow(
            <LmcResidentsSidebar
                residents={residents}
                selectedResident={residents[0].id}
                onCreate={onCreateMock}
            />
        )
    })

    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders an unordered list', () => {
        expect(wrapper.find('ul').length).toEqual(1)
    })

    it('renders a number of sidebar items based on its active residents list', () => {
        expect(wrapper.find('LmcSidebarItem').length).toEqual(2)
    })

    it('renders the sidebar filter', () => {
        expect(wrapper.find('LmcResidentsSidebarFilter').length).toEqual(1)
    })

    it('triggers its onCreate prop', () => {
        wrapper.find('GlyphButton').simulate('click')
        expect(onCreateMock).toBeCalledTimes(1)
        expect(onCreateMock).toBeCalledWith('Residents')
    })
})