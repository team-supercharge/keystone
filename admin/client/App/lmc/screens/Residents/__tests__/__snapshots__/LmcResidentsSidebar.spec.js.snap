// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcResidentsSidebar should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="lmc-sidebar"
    style={
      Object {
        "flex": "1",
        "zIndex": "1",
      }
    }
  >
    <LmcResidentsSidebarFilter
      isChecked={false}
      onFormChange={[Function]}
      onSwitchChange={[Function]}
    />
    <LmcCreateButton
      listId="Resident"
      onCreate={[MockFunction]}
      style={
        Object {
          "borderRadius": 0,
        }
      }
      title="Add a new Resident"
    />
    <ul
      className="lmc-sidebar-list"
    >
      <LmcSidebarItem
        isSelected={true}
        itemData={
          Object {
            "id": "testId1",
            "name": Object {
              "first": "test",
              "last": "id1",
            },
            "status": "active",
          }
        }
        onClick={[Function]}
      />
      <LmcSidebarItem
        isSelected={false}
        itemData={
          Object {
            "id": "testId2",
            "name": Object {
              "first": "test",
              "last": "id2",
            },
            "status": "active",
          }
        }
        onClick={[Function]}
      />
    </ul>
  </div>,
  "nodes": Array [
    <div
      className="lmc-sidebar"
      style={
        Object {
          "flex": "1",
          "zIndex": "1",
        }
      }
    >
      <LmcResidentsSidebarFilter
        isChecked={false}
        onFormChange={[Function]}
        onSwitchChange={[Function]}
      />
      <LmcCreateButton
        listId="Resident"
        onCreate={[MockFunction]}
        style={
          Object {
            "borderRadius": 0,
          }
        }
        title="Add a new Resident"
      />
      <ul
        className="lmc-sidebar-list"
      >
        <LmcSidebarItem
          isSelected={true}
          itemData={
            Object {
              "id": "testId1",
              "name": Object {
                "first": "test",
                "last": "id1",
              },
              "status": "active",
            }
          }
          onClick={[Function]}
        />
        <LmcSidebarItem
          isSelected={false}
          itemData={
            Object {
              "id": "testId2",
              "name": Object {
                "first": "test",
                "last": "id2",
              },
              "status": "active",
            }
          }
          onClick={[Function]}
        />
      </ul>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcResidentsSidebar
        onCreate={[MockFunction]}
        residents={
          Array [
            Object {
              "id": "testId1",
              "name": Object {
                "first": "test",
                "last": "id1",
              },
              "status": "active",
            },
            Object {
              "id": "testId2",
              "name": Object {
                "first": "test",
                "last": "id2",
              },
              "status": "active",
            },
            Object {
              "id": "testId3",
              "name": Object {
                "first": "test",
                "last": "id3",
              },
              "status": "inactive",
            },
          ]
        }
        selectedResident="testId1"
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcResidentsSidebar {
        "_reactInternalInstance": [Circular],
        "calculateHidden": [Function],
        "context": Object {},
        "handleFormChange": [Function],
        "handleSwitchChange": [Function],
        "props": Object {
          "onCreate": [MockFunction],
          "residents": Array [
            Object {
              "id": "testId1",
              "name": Object {
                "first": "test",
                "last": "id1",
              },
              "status": "active",
            },
            Object {
              "id": "testId2",
              "name": Object {
                "first": "test",
                "last": "id2",
              },
              "status": "active",
            },
            Object {
              "id": "testId3",
              "name": Object {
                "first": "test",
                "last": "id3",
              },
              "status": "inactive",
            },
          ],
          "selectedResident": "testId1",
        },
        "refs": Object {},
        "state": Object {
          "displayActiveResidents": true,
          "nameFilter": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="lmc-sidebar"
          style={
            Object {
              "flex": "1",
              "zIndex": "1",
            }
          }
        >
          <LmcResidentsSidebarFilter
            isChecked={false}
            onFormChange={[Function]}
            onSwitchChange={[Function]}
          />
          <LmcCreateButton
            listId="Resident"
            onCreate={[MockFunction]}
            style={
              Object {
                "borderRadius": 0,
              }
            }
            title="Add a new Resident"
          />
          <ul
            className="lmc-sidebar-list"
          >
            <LmcSidebarItem
              isSelected={true}
              itemData={
                Object {
                  "id": "testId1",
                  "name": Object {
                    "first": "test",
                    "last": "id1",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
            <LmcSidebarItem
              isSelected={false}
              itemData={
                Object {
                  "id": "testId2",
                  "name": Object {
                    "first": "test",
                    "last": "id2",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
          </ul>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="lmc-sidebar"
          style={
            Object {
              "flex": "1",
              "zIndex": "1",
            }
          }
        >
          <LmcResidentsSidebarFilter
            isChecked={false}
            onFormChange={[Function]}
            onSwitchChange={[Function]}
          />
          <LmcCreateButton
            listId="Resident"
            onCreate={[MockFunction]}
            style={
              Object {
                "borderRadius": 0,
              }
            }
            title="Add a new Resident"
          />
          <ul
            className="lmc-sidebar-list"
          >
            <LmcSidebarItem
              isSelected={true}
              itemData={
                Object {
                  "id": "testId1",
                  "name": Object {
                    "first": "test",
                    "last": "id1",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
            <LmcSidebarItem
              isSelected={false}
              itemData={
                Object {
                  "id": "testId2",
                  "name": Object {
                    "first": "test",
                    "last": "id2",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
          </ul>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcResidentsSidebar
    onCreate={[MockFunction]}
    residents={
      Array [
        Object {
          "id": "testId1",
          "name": Object {
            "first": "test",
            "last": "id1",
          },
          "status": "active",
        },
        Object {
          "id": "testId2",
          "name": Object {
            "first": "test",
            "last": "id2",
          },
          "status": "active",
        },
        Object {
          "id": "testId3",
          "name": Object {
            "first": "test",
            "last": "id3",
          },
          "status": "inactive",
        },
      ]
    }
    selectedResident="testId1"
  />,
}
`;
